scripts/nbd-firewall-config.sh

```diff
@@ -0,0 +1,91 @@
+#
+##############################################################
+
+# Historical note: the original version of this file filtered
+# by local IP addresses rather than by network interfaces;
+# that version might one day be a useful reference for something.
+
+# Require the "set" argument because otherwise there would be a risk
+# of someone running the script with no arguments and accidentally
+# setting NBD to be available on no addresses.
+# This way, omitting "set" (or supplying "--help") causes the script
+# to print usage instructions.
+
+set -e
+OP="$1"
+if [ _"${OP}" != _set ]; then
+    echo "Usage: $(basename ${0}) set [interface...]
+will set the firewall to allow incoming TCP connections
+to the NBD port on only the specified interfaces (if any)." 1>&2
+    exit 1
+fi
+shift 1
+
+set -eu
+# No automatic cleanup on error because it would depend too
+# critically on the stage at which the error occurs.
+
+NBDPORT=10809
+# Note that chain names must be under 29 characters.
+NEWCHAININ=xapi_nbd_input_chain_new
+NBDCHAININ=xapi_nbd_input_chain
+NEWCHAINOUT=xapi_nbd_output_chain_new
+NBDCHAINOUT=xapi_nbd_output_chain
+
+# Rule fragments for sending NBD packets to NBD chains
+# from INPUT and OUTPUT chains.
+NBD_INPUT_JUMP_TO="-p tcp --dport $NBDPORT -j "
+NBD_OUTPUT_JUMP_TO="-p tcp --sport $NBDPORT -j "
+
+# Same principle as double-buffering a display:
+# add the items one by one to temporary chains,
+# then swap them into use atomically, or as near as possible.
+iptables --new-chain $NEWCHAININ 2>/dev/null || iptables --flush $NEWCHAININ
+iptables --new-chain $NEWCHAINOUT 2>/dev/null || iptables --flush $NEWCHAINOUT
+while [ "$#" -ne 0 ]; do
+    intf="${1}"
+    # (Each positional param is forgotten for ever at the end of the loop.)
+    shift 1
+    iptables -A $NEWCHAININ -m conntrack --ctstate NEW,ESTABLISHED --in-interface "${intf}" -j ACCEPT
+    iptables -A $NEWCHAINOUT --out-interface "${intf}" -j RETURN
+done
+iptables --append $NEWCHAININ -j REJECT
+iptables --append $NEWCHAINOUT -j REJECT
+
+function insert_rule_if_absent {
+    chain="${1}"
+    shift 1
+    rule="${*}"
+    if ! iptables --check "${chain}" $rule 2>/dev/null
+    then
+        iptables --insert "${chain}" $rule
+    fi
+}
+
+# Start using the new chains.
+insert_rule_if_absent INPUT "${NBD_INPUT_JUMP_TO} ${NEWCHAININ}"
+insert_rule_if_absent OUTPUT "${NBD_OUTPUT_JUMP_TO} ${NEWCHAINOUT}"
+
+# If old chains are present, stop using them and then delete them.
+iptables --delete INPUT $NBD_INPUT_JUMP_TO $NBDCHAININ 2>/dev/null || true
+iptables --flush "${NBDCHAININ}" 2>/dev/null && \
+    iptables --delete-chain "${NBDCHAININ}"
+
+iptables --delete OUTPUT $NBD_OUTPUT_JUMP_TO $NBDCHAINOUT 2>/dev/null || true
+iptables --flush "${NBDCHAINOUT}" 2>/dev/null && \
+    iptables --delete-chain "${NBDCHAINOUT}"
+
+# Rename our rules so they no longer appear new. (References remain valid.)
+iptables --rename-chain "${NEWCHAININ}" "${NBDCHAININ}"
```

Comment:
Is it required to rename NEW to NBD. iptables rules might still pointing to NEW, which might not exist after rename. 
Why cant we directly append rules to NBD chain ?

Reply:
The rename is cosmetic: any reference to it remains valid.
A demonstration:
```
[root@xrtuk-01-01 ~]# iptables --list-rules demochain
-N demochain
-A demochain -j demotwo
[root@xrtuk-01-01 ~]# iptables --list demochain
Chain demochain (0 references)
target     prot opt source               destination         
demotwo    all  --  anywhere             anywhere            
[root@xrtuk-01-01 ~]# iptables --rename-chain demotwo demothree
[root@xrtuk-01-01 ~]# iptables --list-rules demochain
-N demochain
-A demochain -j demothree
[root@xrtuk-01-01 ~]# iptables --list demochain
Chain demochain (0 references)
target     prot opt source               destination         
demothree  all  --  anywhere             anywhere
```

Reply:
Ok, --rename-chain renames the rules as well, that is good enough. But why rename ? it doesn't seem to add any value here.

Reply:
So, I've shown why it's safe that way.
As for the other approach, that would mean making changes one rule at a time to a rule while it is in use. That way we would need to examine or rely on the initial contents of the rule, and be very careful about the order of operations.

I felt it was safer and cleaner to make a new rule and populate it correctly, then switch over to using it.
As the comment says,
```
# Same principle as double-buffering a display:
# add the items one by one to temporary chains,
# then swap them into use atomically, or as near as possible.
```

Reply:
I'm doing it for the cases when the script has been run before with different arguments, so there is already a `xapi_nbd_input_chain` in use (and output equivalent).

I suppose the other way to achieve broadly the same thing would have been to start by renaming the pre-existing `xapi_nbd_input_chain` (if present) to `xapi_nbd_input_chain_soon_to_be_replaced` or something, then make a new `xapi_nbd_nbd_input`, then delete the old one.

Reply:
lgtm